### ISSUE TITLE
Type checker can be made to consume all available memory

### ISSUE THREAD

**Issue by crazymykl** [elm-lang/elm-compiler#1380]
I discovered this by accident due to a typo in some more involved code, but  I believe I've boiled it down to a minimal repro case: 

```elm
import Array exposing (Array)

fribble x y =
  Array.get y x
    |> Maybe.map (Array.set y x)
    |> Maybe.withDefault x
```

When I try to compile this code, `elm-make` hangs, and hitting `^C` does not shutdown all elm-make processes, meaning memory usage grows rapidly until I or the OOM killer address the situation.

**Comment by jvoigtlaender** [comment]
Looks like another instance of "infinite type construction not caught by the compiler", see https://github.com/elm-lang/elm-compiler/issues/1231 and other issues referenced there and referencing that one.

**Comment by auxsoft** [comment]
This still crashes the compiler in version 0.18.
